File: src/mentoai/ai/recommend.py
import logging

from fastapi import HTTPException

from mentoai.ai.embeddings import embed_query
from mentoai.ai.schemas import JobSummary, RecommendationListResponse
from mentoai.ai.scoring import build_reason, similarity_to_score, skill_overlap
from mentoai.ai.users import build_profile_text, fetch_user_info
from mentoai.config import get_settings
from mentoai.db.pool import fetch

logger = logging.getLogger(__name__)

SEARCH_SQL = """
SELECT j.id, j.source, j.company, j.position, j.skill_tags, j.annual_from, j.annual_to,
       j.is_newbie, j.location,
       1 - (e.embedding <=> $1) AS similarity
FROM silver.job_embeddings e
JOIN silver.jobs j ON j.id = e.job_id
ORDER BY e.embedding <=> $1
LIMIT $2
"""


def career_label(annual_from: int | None, annual_to: int | None, is_newbie: bool | None) -> str | None:
    if is_newbie:
        return "신입 가능"
    if annual_from is not None:
        if (annual_to or 0) >= 99:  # Wanted의 '상한 없음'은 100으로 온다
            return f"경력 {annual_from}년 이상"
        return f"경력 {annual_from}~{annual_to or annual_from}년"
    return None


async def recommend_jobs(user_id: int) -> RecommendationListResponse:
    """벡터 유사도 + 휴리스틱으로 추천. LLM 호출/비용 없이 즉시 응답한다."""
    try:
        user = await fetch_user_info(user_id)
        query_vector = await embed_query(build_profile_text(user))
        rows = await fetch(SEARCH_SQL, query_vector, get_settings().recommend_top_k)
    except HTTPException:
        raise
    except Exception as error:
        logger.exception("추천 조회 실패 (user_id=%s)", user_id)
        raise HTTPException(status_code=500, detail=str(error)) from error

    user_skills = list(user["skills"] or [])
    career_years = user["career_years"]
    recommendations = [
        JobSummary(
            job_id=row["id"],
            company=row["company"] or "미상",
            title=row["position"] or "미상",
            source=row["source"],
            career=career_label(row["annual_from"], row["annual_to"], row["is_newbie"]),
            location=row["location"],
            skills=list(row["skill_tags"] or [])[:4],
            matched_skills=skill_overlap(user_skills, list(row["skill_tags"] or [])),
            match_score=similarity_to_score(row["similarity"]),
            reason=build_reason(user_skills, career_years, dict(row)),
        )
        for row in rows
    ]
    return RecommendationListResponse(
        user_name=user["username"], recommendations=recommendations
    )

File: internal/recommend/recommend.go
// Package recommend은 벡터 유사도 + 휴리스틱으로 채용공고를 추천한다.
// LLM 호출이 없어 즉시 응답한다.
package recommend

import (
	"context"
	"fmt"
	"log/slog"
	"strings"

	"github.com/Chae-JS/mentoai/internal/domain"
	"github.com/Chae-JS/mentoai/internal/embedding"
	"github.com/Chae-JS/mentoai/internal/scoring"
	"github.com/Chae-JS/mentoai/internal/storage"
)

// EmbedderProvider는 현재 활성 임베딩 구현을 제공한다.
// embedding.Embedder는 검색에 필요한 메서드만 노출하는 좁은 포트로 쓰인다.
type EmbedderProvider interface {
	Current() embedding.Embedder
}

// SettingsProvider는 추천 top_k 같은 런타임 설정을 읽는다.
type SettingsProvider interface {
	RecommendTopK() int
}

// Service는 추천 서비스다.
type Service struct {
	users    storage.UserRepo
	embeds   storage.EmbeddingRepo
	jobs     storage.JobRepo
	embedder EmbedderProvider
	settings SettingsProvider
}

// New는 의존성을 주입받아 추천 서비스를 만든다.
func New(users storage.UserRepo, embeds storage.EmbeddingRepo, jobs storage.JobRepo, embedder EmbedderProvider, settings SettingsProvider) *Service {
	return &Service{users: users, embeds: embeds, jobs: jobs, embedder: embedder, settings: settings}
}

// BuildProfileText는 사용자 스펙을 임베딩 쿼리 텍스트로 만든다.
func BuildProfileText(user domain.UserInfo) string {
	return fmt.Sprintf("희망직무: %s, 보유기술: %s, 경력: %d년",
		user.DesiredJob, strings.Join(user.Skills, ", "), user.CareerYears)
}

// Recommend는 사용자 프로필과 유사한 공고 상위 k건을 돌려준다.
func (s *Service) Recommend(ctx context.Context, userID int64) (domain.RecommendationListResponse, error) {
	var out domain.RecommendationListResponse

	user, err := s.users.Info(ctx, userID)
	if err != nil {
		return out, err
	}
	if user == nil {
		return out, domain.NotFound("User not found")
	}

	queryVector, err := s.embedder.Current().EmbedQuery(ctx, BuildProfileText(*user))
	if err != nil {
		slog.Error("추천 조회 실패", "user_id", userID, "error", err)
		return out, domain.AsHTTPError(err)
	}

	hits, err := s.embeds.SearchTopK(ctx, queryVector, s.settings.RecommendTopK())
	if err != nil {
		slog.Error("추천 조회 실패", "user_id", userID, "error", err)
		return out, domain.AsHTTPError(err)
	}

	ids := make([]int64, len(hits))
	for i, h := range hits {
		ids[i] = h.JobID
	}
	metas, err := s.jobs.GetMetaByIDs(ctx, ids)
	if err != nil {
		slog.Error("추천 조회 실패", "user_id", userID, "error", err)
		return out, domain.AsHTTPError(err)
	}

	recommendations := []domain.JobSummary{}
	for _, hit := range hits {
		meta, ok := metas[hit.JobID]
		if !ok {
			continue // 임베딩은 있지만 공고가 삭제된 경우
		}
		recommendations = append(recommendations, toJobSummary(hit, meta, user.Skills, user.CareerYears))
	}
	return domain.RecommendationListResponse{UserName: user.Username, Recommendations: recommendations}, nil
}

func toJobSummary(hit domain.SearchHit, meta domain.JobMeta, userSkills []string, careerYears int) domain.JobSummary {
	skills := meta.SkillTags
	if len(skills) > 4 {
		skills = skills[:4]
	}
	return domain.JobSummary{
		JobID:         meta.ID,
		Company:       defaultStr(meta.Company, "미상"),
		Title:         defaultStr(meta.Position, "미상"),
		Source:        meta.Source,
		Career:        scoring.CareerLabel(meta.AnnualFrom, meta.AnnualTo, meta.IsNewbie),
		Location:      meta.Location,
		Skills:        skills,
		MatchedSkills: scoring.SkillOverlap(userSkills, meta.SkillTags),
		MatchScore:    scoring.SimilarityToScore(hit.Similarity),
		MaxScore:      100,
		Reason:        scoring.BuildReason(userSkills, careerYears, meta.SkillTags, meta.AnnualFrom, meta.IsNewbie),
	}
}

func defaultStr(s *string, def string) string {
	if s == nil || *s == "" {
		return def
	}
	return *s
}
